node/bft/src/helpers/pending.rs

```diff
@@ -146,29 +160,55 @@ impl<T: Copy + Clone + PartialEq + Eq + Hash, V: Clone> Pending<T, V> {
     /// If the `item` does not exist, `None` is returned.
     pub fn remove(&self, item: impl Into<T>, callback_value: Option<V>) -> Option<HashSet<SocketAddr>> {
         let item = item.into();
-        // Remove the item from the pending queue.
-        let result = self.pending.write().remove(&item);
-        // Remove the callback for the item, and process any remaining callbacks.
-        if let Some(callbacks) = self.callbacks.lock().remove(&item) {
-            if let Some(callback_value) = callback_value {
-                // Send a notification to the callback.
-                for (callback, _, _) in callbacks {
-                    callback.send(callback_value.clone()).ok();
+        // Remove the item from the pending queue and process any remaining callbacks.
+        match self.pending.write().remove(&item) {
+            Some(callbacks) => {
+                // Get the peer IPs.
+                let peer_ips = callbacks.keys().cloned().collect();
+                // Process the callbacks.
+                if let Some(callback_value) = callback_value {
+                    // Send a notification to the callback.
+                    for (callback, _, _) in callbacks.into_values().flat_map(|callbacks| callbacks.into_iter()) {
+                        callback.send(callback_value.clone()).ok();
+                    }
                 }
+                // Return the peer IPs.
+                Some(peer_ips)
             }
+            None => None,
         }
-        // Return the result.
-        result
     }
 
     /// Removes the callbacks for the specified `item` that have expired.
-    pub fn clear_expired_callbacks_for_item(&self, item: impl Into<T>) {
+    pub fn clear_expired_callbacks_for_item(&self, now: i64, item: impl Into<T>) {
+        let item = item.into();
+
+        // Acquire the pending lock.
+        let mut pending = self.pending.write();
```

Comment:
this lock is acquired in a loop; it might make sense to tweak the API so that it's done just once for the duration of the `clear_expired_callbacks` loop

Reply:
Resolved in https://github.com/AleoHQ/snarkOS/pull/3135/commits/964599be38d5dfe56f5d4701ad4cccf98819c8d1. 

This also removes a use of `collect`. 

@ljedrz Can you review the new change?

Reply:
Looks good :+1:; we can now consider if we still need the `clear_expired_callbacks_for_item` method.

Reply:
We still use the `clear_expired_callbacks_for_item` on individual calls to `num_sent_requests`, `num_callbacks`, and `insert`. Primarily to clear the expired callbacks in order to fetch the correct output number.